### ISSUE TITLE
Dependency cookbooks aren't being extracted by `install` command on Windows

### ISSUE THREAD

**Issue by kenny-evitt** [hw-labs/batali#62]
I've got a Chef organization repo and its *Batali* file contains only the following:

```
Batali.define do
  source 'https://supermarket.chef.io'

  cookbook 'myapp', git: 'https://user:password@mygitserver.example.com/myapp-chef-cookbook.git'
end
```

Running the `resolve` command produces a *batali.manifest* that seems correct:

```
{
  "infrastructure": false,
  "cookbook": [
    {
      "name": "myapp",
      "dependencies": [
        [
          "chocolatey",
          "> 0"
        ]
      ],
      "version": "0.1.0",
      "source": {
        "url": "https://user:password@mygitserver.example.com/myapp-chef-cookbook.git",
        "ref": "blah-blah-blah",
        "type": "Batali::Source::Git",
        "subdirectory": null
      }
    },
    {
      "name": "chocolatey",
      "dependencies": [
        [
          "windows",
          "~> 1.38"
        ]
      ],
      "version": "0.6.1",
      "source": {
        "type": "Batali::Source::Site",
        "url": "https://supermarket.chef.io:443/api/v1/cookbooks/chocolatey/versions/0.6.1/download",
        "version": "0.6.1"
      }
    },
    {
      "name": "windows",
      "dependencies": [
        [
          "chef_handler",
          ">= 0.0.0"
        ]
      ],
      "version": "1.38.4",
      "source": {
        "type": "Batali::Source::Site",
        "url": "https://supermarket.chef.io:443/api/v1/cookbooks/windows/versions/1.38.4/download",
        "version": "1.38.4"
      }
    },
    {
      "name": "chef_handler",
      "dependencies": [

      ],
      "version": "1.2.0",
      "source": {
        "type": "Batali::Source::Site",
        "url": "https://supermarket.chef.io:443/api/v1/cookbooks/chef_handler/versions/1.2.0/download",
        "version": "1.2.0"
      }
    }
  ]
}
```

With the changes for #61, running `install` I get a sub-directory *myapp* under the *cookbooks* directory but instead of sub-directories for the dependency cookbooks I just see the 'tar.gz' files for those cookbooks.

I'm looking into this now. I suspect more Windows-specific problems like #60.

**Comment by chrisroberts** [comment]
Just to confirm the behavior, would you run:

```
$ batali cache --scrub
```
and then run the install again to ensure that the issue isn't resulting from the state of the local cache. Thanks!

**Comment by kenny-evitt** [comment]
@chrisroberts I was, probably, doing the same thing as `batali cache --scrub` – I was deleting the contents of the *~/.batali/cache* sub-directories *git* and *site*.

But I ran `chef exec batali cache --scrub` and then `chef exec batali install` again.

Now I don't see anything under my org-repo *cookbooks* sub-directory. And I'm getting the same error as in #60.

I compared the *asset* file in the 'random' *site* sub-directory with the corresponding file in the *site/entitystore* sub-directory and the *asset* file is missing two bytes – at bytes 2048 and 4096. I cannot for the life of me figure out why those bytes are missing. There doesn't seem to be anything special about those bytes other than their positions – the values are `C9` and `72`.

Very strange!

**Comment by chrisroberts** [comment]
Can you properly unpack the file that is located within the entitystore?

**Comment by kenny-evitt** [comment]
@chrisroberts Yes.

**Comment by chrisroberts** [comment]
@kenny-evitt could you point your bundle at the branch on that PR (#63 - `fix/windows-caching`) and see if that resolves things. May need to re-scrub your cache as well. Thanks!

**Comment by kenny-evitt** [comment]
@chrisroberts That helped! But now I've got a new error. It seems like the random sub-directory for the *asset* files are being created as read-only.

I commented out the line in *site.rb* that deletes the *asset* files but I think doing so results in the unexpected behavior I originally reported for this issue – the tar-gz files (i.e. the *asset* files) are being copied to my org repo's *cookbooks* directory.

Adding a `:mode` options argument to the `mkdir_p` calls don't seem to help.

**Comment by chrisroberts** [comment]
Did you scrub the cache prior to running with the `:mode` change? It won't be applied if the directory is already there. Failing that, might need to do an explicit `File.chmod` to set it as expected.

**Comment by kenny-evitt** [comment]
@chrisroberts I was manually deleting the *cache* directory and it still wasn't working. `FileUtils.chmod` also does not seem to be working.

I wonder if running everything with `chef exec ...` might be part of the problem. Troubleshooting Chef on Windows is dispiriting.

**Comment by kenny-evitt** [comment]
A similar issue reported for a Ruby file upload library:

 - [Permission denied @ unlink_internal · Issue #133 · refile/refile](https://github.com/refile/refile/issues/133)

**Comment by chrisroberts** [comment]
Is the parent directory of the cache on a networked file system?

**Comment by kenny-evitt** [comment]
@chrisroberts Yeah. That could certainly be the culprit. I noticed that some other directories/files on the same filesystem (accessed via a 'mapped drive' pointing to the relevant 'network share') have the same read-only property set.

I'm working on a workaround that I don't expect would make sense to merge – the idea being that I'll tweak the code at the end of the `asset` method in *site.rb* to return the extracted sub-directory for the dependency cookbooks, instead of the *asset* file as it does now.

**Comment by kenny-evitt** [comment]
@chrisroberts My workaround seems to be working. I've got the dependency cookbook directories under my org-repo *cookbooks* sub-directory now. The change consists of removing [this line](https://github.com/hw-labs/batali/blob/fix/windows-caching/lib/batali/source/site.rb#L83) [I'm referencing the version as-of your recent PR #63]:

```
            FileUtils.rm(a_path)
```

and changing [this line](https://github.com/hw-labs/batali/blob/fix/windows-caching/lib/batali/source/site.rb#L94):

```
        Dir.glob(File.join(path, '*')).first
```

to:

```
        Dir.glob(File.join(path, '*')).reject { |m| m.end_with? '/asset' }.first
```

**Comment by chrisroberts** [comment]
@kenny-evitt could you give one thing a quick try just to see if it may be an alternate solution. Updating this line: 

https://github.com/hw-labs/batali/blob/fix/windows-caching/lib/batali/source/site.rb#L64

to be:

```ruby
File.open(a_path = File.join(path, 'asset.tgz'), 'wb') do |file|
```

I've seen some references to the perms being weird if there is no extension provided on the file, and am wondering if that might be the problem. Either way, at least a fix is available \o/

Thanks!

**Comment by kenny-evitt** [comment]
@chrisroberts I tried exactly that, tho I opted to use a dummy extension – `.blah` – instead. It didn't work.

I'm having other problems now tho, like paths containing a mix of standard Windows path separators and Unix-style separators (that are actually appropriate when running Chef under MinGW), so I'm going to try uninstalling and re-installing the Chef DK on my computer next.

I'll update this with more info when I can test again.

Thanks for your help!

**Comment by chrisroberts** [comment]
@kenny-evitt awesome, thanks so much for all the feedback. I'll try and get the path separators swapped out to use the constant in `File` before you circle back around to this again. Cheers!

**Comment by kenny-evitt** [comment]
@chrisroberts Sorry! I'm not having other problems with Batali – I'm having problems with Chef itself. I'll create a new issue if I run into any other problems with Batali. But, with my minor tweak, everything seems to be working perfectly.

**Comment by chrisroberts** [comment]
Oh, okay great! I'll get the change integrated and released as soon as I can. Will update this issue and close once it's out. Thanks!

**Comment by webframp** [comment]
Working for me. Also tested with chefdk with explicit: `gem build` and `gem install batali-0.3.5.gem --install-dir C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0 --no-user-install`

```
C:\src\chef\infra-repo [master +12 ~3 -0 !]> batali update
[Batali]: Loading sources... complete!
[Batali]: Performing single path resolution.
[Batali]: Resolving dependency constraints... complete!
[Batali]: Writing manifest... complete!
[Batali]: Ideal solution:
chef_handler <1.2.0>
iis <4.1.5>
windows <1.38.4>
[Batali]: Readying installation destination... complete!
[Batali]: Installing cookbooks... complete!
```

```
C:\src\chef\infra-repo [master +13 ~3 -0 !]> batali --version
batali - Cookbook Collection Manager - [Version: 0.3.5]
  - I got 99 problems but a policyfile ain't one
```

**Comment by chrisroberts** [comment]
Fixed and released in v0.3.6. Thanks so much for the help @kenny-evitt and @webframp !
